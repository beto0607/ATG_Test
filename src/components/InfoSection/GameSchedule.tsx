import React, { useState } from "react";
const GameSchedule: React.FC = ({ title }) => {
    return (
        <div>
            <h4>{title}</h4>
        </div>
    );
};

export default GameSchedule;
