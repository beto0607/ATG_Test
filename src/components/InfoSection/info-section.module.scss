$box-border: 1px solid #666;
.day {
    display: block;
    width: fit-content;
    padding: 4px 10px;
    font-weight: bold;
    text-decoration: underline;
}
.list-group {
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    justify-content: space-between;
    justify-items: left;
    .list-group-item {
        display: block;
        flex-basis: 100%;
        margin: 10px auto;
        text-align: center;
        border-color: transparent;
    }
}
.game-info-day-container {
    border: $box-border;
    padding: 20px;
    margin: 10px auto;
}
