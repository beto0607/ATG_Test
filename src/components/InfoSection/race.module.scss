.start-container {
    // border: 1px solid #ccc;
    margin: 10px auto;
    .start-title-container {
        cursor: pointer;
        padding: 10px 20px;
    }
    span {
        display: block;
        width: fit-content;
        margin: auto;
    }
}
.race-container {
    border: 1px solid #ccc;
    .race-title-container {
        margin: 20px auto;
    }
}
