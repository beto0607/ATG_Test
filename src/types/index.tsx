type TrackInfo = {
    id: number,
    name: string
};
type FavoriteInfo = {
    raceId: string,
    legNumber: number,
    start: {
        number: number,
        name: string,
        distribution: number
    }
};
type GameInfo = {
    id: string,
    startTime: string, // YYYY-MM-DDThh:mm:ss
    tracks: Array<TrackInfo>,
    favorites: Array<FavoriteInfo>
}
//ENDPOINT RESPONSE
//https://www.atg.se/services/racinginfo/v1/api/products/<gameType>
type GameSchedule = {
    betType: string,
    upcoming: Array<GameInfo>,
    results: Array<GameInfo>
};

type RaceStart = {
    number: number,
    driver: {
        firstName: string,
        lastName: string
    },
    horse: {
        name: string,
        trainer: {
            firstName: string,
            lastName: string
        },
        pedigree: {
            father: {
                name: string
            }
        }
    }
}
type GameRace = {
    date: string,
    name: string,
    scheduledStartTime: string,
    starts: Array<RaceStart>
}
//ENDPOINT RESPONSE
// https://www.atg.se/services/racinginfo/v1/api/games/<gameId>
type GameResponse = {
    id: string,
    status: string,
    races: Array<GameRace>
};

export interface ApplicationState {
    text: string;
    game_schedule: GameSchedule;
    game_data: GameResponse;
}