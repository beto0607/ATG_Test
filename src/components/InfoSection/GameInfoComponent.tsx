import React, { useState } from "react";
import { GameInfo } from '../../types/index';

const GameInfoComponent: React.FC<GameInfo> = ({ id, startTime, tracks, favorites }: GameInfo) => {
    return (
        <li>
            <span>#{id}</span>
        </li>
    );
};

export default GameInfoComponent;
