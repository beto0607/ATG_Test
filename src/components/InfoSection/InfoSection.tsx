import React from 'react';
import styles from './info-section.module.scss';
import { connect } from "react-redux";
import GameSchedule from './GameSchedule';
import { GameSchedule as GameScheduleType } from '../../types/index';
import { Container, Row, Col, Alert } from "react-bootstrap";
import { ApplicationState } from '../../types/index';

//Title component for isolate logic
export interface TitleProps {
    text: string;
    betType: string;
};
export const Title: React.FC<TitleProps> = ({ text, betType }: TitleProps) => (
    <Alert variant={!text ? 'info' : (!betType ? 'danger' : 'success')}>
        <h2>{!text ? 'You can search for a Game now' : (betType ? `Bet type: ${betType}` : `Game "${text}" not found`)}</h2>
    </Alert>
);
// Info section
export interface ConnectedInfoSectionProps {
    text: string;
    gameSchedule: GameScheduleType;
};
export const ConnectedInfoSection: React.FC<ConnectedInfoSectionProps> = ({ text, gameSchedule }: ConnectedInfoSectionProps) => (
    <Container className={styles['info-section']}>
        <Row>
            <Col>
                <Title text={text} betType={gameSchedule.betType || ''} />
            </Col>
        </Row>
        <Row>
            <Col xs={12} md={6}>
                {gameSchedule.upcoming.length && <GameSchedule title={"Closests upcomings"} games={gameSchedule.upcoming || []} />}
            </Col>
            <Col xs={12} md={6}>
                {gameSchedule.results.length && <GameSchedule title={"Closets results"} games={gameSchedule.results || []} />}
            </Col>
        </Row>
    </Container>
);
const mapStateToProps = (state: ApplicationState) => {
    return {
        text: state.text,
        gameSchedule: state.gameSchedule
    };
}
const InfoSection = connect(mapStateToProps)(ConnectedInfoSection);
export default InfoSection;
