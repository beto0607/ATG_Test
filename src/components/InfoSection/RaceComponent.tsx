import React, { useState } from "react";
import { Collapse, Badge, Alert } from "react-bootstrap";
import { Race, RaceStart } from '../../types/index';
import Moment from "react-moment";
import styles from './race.module.scss';

export const StartComponent: React.FC<RaceStart> = ({ number, horse, driver }: RaceStart) => {
    const [open, setOpen] = useState(false);
    return (
        <div onClick={() => setOpen(!open)} className={styles['start-container']}>
            <div className={styles['start-title-container']}>
                <Badge variant="info">Start: #{number}</Badge>
                <span>Horse: <strong>{horse.name}</strong></span>
                <span>Driver: <strong>{`${driver.firstName} ${driver.lastName}`}</strong></span>
            </div>
            <Collapse in={open}>
                <div>
                    <span>Trainer: {`${horse.trainer.firstName} ${horse.trainer.lastName}`}</span>
                    <span>Horse father: {`${horse.pedigree.father.name}`}</span>
                </div>
            </Collapse>
        </div>
    )
};
export const RaceComponent: React.FC<Race> = ({ id, name, number, scheduledStartTime, starts }: Race) => {
    return (
        <div className={styles['race-container']}>
            <div className={styles['race-title-container']}>
                <Alert variant='info'>
                    <span>#{number} {name && <strong>- {name}</strong>}</span>
                </Alert>
                <Alert variant='info'>
                    Start time:
                    <Moment date={scheduledStartTime} format='HH:mm:ss' />
                </Alert>
            </div>
            {starts.map(start => <StartComponent key={`${id}_${start.number}`} {...start} />)}
        </div>
    );
};