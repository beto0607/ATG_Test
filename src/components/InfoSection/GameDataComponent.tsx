import React from "react";
import { connect } from "react-redux";
import { loadGameData } from "../../actions/actions";
import { ApplicationState, GameData, Race } from '../../types/index';
import { RaceComponent } from './RaceComponent';

export interface GameDataProps {
    parent_id: string;
    gameData: GameData | null;
    loadGameData: any;
    load: boolean;
}
export const GameDataComponentConnected: React.FC<GameDataProps> = ({ parent_id, load, gameData, loadGameData }: GameDataProps) => {
    // Loads the game data only when the component is visible and only one time.
    if (load && !gameData) { loadGameData(parent_id); }
    return (
        <div>
            {gameData !== null && gameData.races.map((race: Race) => <RaceComponent key={race.id} {...race} />)}
        </div>
    );
};
const mapStateToProps = (state: ApplicationState, ownprops: any): GameData | null => {
    return {
        ...ownprops,
        gameData: state.gameData.find((e: GameData) => e.id === ownprops.parent_id) || null
    }
};
const mapDispatchToPropsGameData = (dispatch: any, ownprops: any): any => {
    return {
        ...ownprops,
        loadGameData: (gameId: string) => dispatch(loadGameData(gameId))
    };
}
export const GameDataComponent = connect(mapStateToProps, mapDispatchToPropsGameData)(GameDataComponentConnected);
