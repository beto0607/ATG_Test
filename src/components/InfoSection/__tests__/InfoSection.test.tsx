import React from 'react';
import { shallow } from 'enzyme';
import { ConnectedInfoSection, ConnectedInfoSectionProps, Title, TitleProps } from '../InfoSection';
import toJson from 'enzyme-to-json'
import { Alert } from 'react-bootstrap';
import { GameInfo } from '../../../types';
import GameSchedule from '../GameSchedule';

describe('Title', () => {
    it('renders without crashing', () => {
        const props: TitleProps = { text: '', betType: '' }
        const component = shallow((<Title {...props} />));
        expect(toJson(component)).toMatchSnapshot()
    });
    it('should show info Alert when text it\'s an empty string', () => {
        const props: TitleProps = { text: '', betType: '' }
        const component = shallow((<Title {...props} />));
        expect(component.find(Alert).prop('variant')).toEqual('info');
    });
    it('should show danger Alert when betType it\'s an empty string', () => {
        const props: TitleProps = { text: 'test', betType: '' }
        const component = shallow((<Title {...props} />));
        expect(component.find(Alert).prop('variant')).toEqual('danger');
    });
    it('should show success Alert when betType it\'s a string', () => {
        const props: TitleProps = { text: 'test', betType: 'test' }
        const component = shallow((<Title {...props} />));
        expect(component.find(Alert).prop('variant')).toEqual('success');
    });
});

const mockGameInfo: GameInfo = {
    id: 'an_id',
    startTime: 'a_start_time',
    tracks: [],
    favorites: []
};
describe('ConnectedInfoSection', () => {
    it('renders without crashing', () => {
        const props: ConnectedInfoSectionProps = {
            text: '',
            gameSchedule: {
                betType: '',
                upcoming: [],
                results: []
            }
        }
        const component = shallow((<ConnectedInfoSection {...props} />));
        expect(toJson(component)).toMatchSnapshot()
    });
    it('renders 1 upcoming GameComponent', () => {
        const props: ConnectedInfoSectionProps = {
            text: '',
            gameSchedule: {
                betType: '',
                upcoming: [mockGameInfo],
                results: []
            }
        }
        const component = shallow((<ConnectedInfoSection {...props} />));
        expect(component.find(GameSchedule)).toHaveLength(1)
    });
    it('renders 1 result GameComponent', () => {
        const props: ConnectedInfoSectionProps = {
            text: '',
            gameSchedule: {
                betType: '',
                upcoming: [],
                results: [mockGameInfo]
            }
        }
        const component = shallow((<ConnectedInfoSection {...props} />));
        expect(component.find(GameSchedule)).toHaveLength(1)
    });
});

