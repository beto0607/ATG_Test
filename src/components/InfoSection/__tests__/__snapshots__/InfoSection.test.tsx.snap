// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConnectedInfoSection renders without crashing 1`] = `
<Container
  className="info-section"
  fluid={false}
>
  <Bootstrap(Row)>
    <Col>
      <Title
        betType=""
        text=""
      />
    </Col>
  </Bootstrap(Row)>
  <Bootstrap(Row)>
    <Col
      md={6}
      xs={12}
    >
      0
    </Col>
    <Col
      md={6}
      xs={12}
    >
      0
    </Col>
  </Bootstrap(Row)>
</Container>
`;

exports[`Title renders without crashing 1`] = `
<Alert
  closeLabel="Close alert"
  show={true}
  transition={[Function]}
  variant="info"
>
  <h2>
    You can search for a Game now
  </h2>
</Alert>
`;
